File: project-euler/problem004.hs
#!/usr/bin/env runhaskell

{-|
  = Problem #4: Largest palindrome product

  A palindromic number reads the same both ways. The largest
  palindrome made from the product of two 2-digit numbers is
  9009 = 91 × 99.

  Find the largest palindrome made from the product of two
  3-digit numbers.
-}

import Test.HUnit


solve :: [Int] -> Int
solve factorsRange = maximum palindromes
    where
        products = [x * y | x <- factorsRange, y <- [x..(last factorsRange)]]
        palindromes = filter palindrome products

palindrome :: Int -> Bool
palindrome n = chars == reverse chars
    where chars = show n

-- tests

testExample = TestCase (assertEqual "As an example," 9009 (solve [10..99]))
testSolution = TestCase (assertEqual "As the solution," 906609 (solve [100..999]))

main = runTestTT (TestList [testExample, testSolution])

File: project-euler/problem004.jl
#!/usr/bin/env julia

# Problem #4: Largest palindrome product
# ======================================
#
# A palindromic number reads the same both ways. The largest
# palindrome made from the product of two 2-digit numbers is
# 9009 = 91 × 99.
#
# Find the largest palindrome made from the product of two
# 3-digit numbers.

using Base.Test

using Iterators


function solve(factorsrange)
    factors = product(factorsrange, factorsrange)
    products = [x[1] * x[2] for x in factors]
    palindromes = filter(palindrome, products)
    maximum(palindromes)
end

function palindrome(n)
    chars = string(n)
    chars == reverse(chars)
end


# --- tests ---

test_example() = @test solve(10:99) == 9009
test_solution() = @test solve(100:999) == 906609

function run_tests()
    test_example()
    test_solution()
    println("Tests were successful.")
end

run_tests()
